### DISCUSSION TITLE
OIndexManager is marked as deprecated, what should be used to interact with existing indexes instead

### DISCUSSION THREAD

**Discussion by scrabb** [orientechnologies/orientdb#9715] [Q&A]
### OS: Windows 10  

OIndexManager is now marked as deprecated, as is getIndexManager() from OMetadata.

What are correct methods for performing tasks on indexes now that these are not to be used?

**Comment by VincentSleepless** [comment]
u can get OIndex obj and query index value like this:

`

    private static void lookUpvertex() {
        logger.info("start to lookup orient vertex class");
        OrientDB orient = null;
        ODatabaseDocumentInternal dbsession = null;
        try {
            orient = new OrientDB("remote:xxxx", OrientDBConfig.defaultConfig());
            dbsession = (ODatabaseDocumentInternal)orient.open("xxx", "xxx", "xxx");
            dbsession.begin(OTransaction.TXTYPE.OPTIMISTIC);
            //OIndex index = dbsession.getMetadata().getIndexManager().getIndex("ZA_TEST_V.id");
            OIndex index = dbsession.getMetadata().getIndexManagerInternal().getIndex(dbsession,"ZA_TEST_V.id");
            logger.info(index.getName());
            OType idxFieldType = index.getDefinition().getTypes()[0];
            Object joinValue = OType.convert("ZA000002", idxFieldType.getDefaultJavaType());
            String queryGet = String.format("select rid from index:`%s` where key = ?","ZA_TEST_V.id");
            OResultSet oResultSet =  dbsession.indexQuery("ZA_TEST_V.id", queryGet, joinValue );
            if (oResultSet != null && oResultSet.hasNext()){
                OIdentifiable oIdentifiable = oResultSet.next().getProperty("rid");
                logger.info(StringUtil.getBeautifyJson(oIdentifiable.getIdentity().toString()));
             }
            dbsession.commit();
        } catch (Exception ex) {
            logger.error("lookup edge vertex  failed ! start to row back", ex);
            dbsession.reload();
            dbsession.rollback();
        }finally {
            orient.close();
        }
        logger.info("finish to lookup orient vertex class");
    }

`

**Comment by andrii0lomakin** [comment]
Hi, internal classes are only for internal usage, they are not guaranteed to be stable. The correct way to use indexes right now is to create a class with indexed field and use documents to store data.

**Comment by VincentSleepless** [comment]
> Hi, internal classes are only for internal usage, they are not guaranteed to be stable. The correct way to use indexes right now is to create a class with indexed field and use documents to store data.

index manager may very useful for developers to write customer etl tools , especially for edge integration lookup index @rid, could you consider to give some stable api for query index by name  or query data by index?

**Comment by andrii0lomakin** [comment]
What stops you to use query which internally will work with the same index?

**Comment by VincentSleepless** [comment]
> What stops you to use query which internally will work with the same index?

i copy the etl tool lookup transformer ~ it use the internal method~

by the way , your suggestion is good~~ i will change validate index from internal to class itself  and change query index value from internal to index query sql ~~~

**Comment by mmacfadden** [comment]
@laa 

Greetings. I would also add that we have used the IndexManager class extensively when creating and dropping indexes. We use the Java API for managing schema rather than SQL.  Meaning we create OClass, OProperty, etc. using the Java API.  We use the index manager in the same way to create indices. If this is deprecated, what Java API should we use?

Indexes are (somewhat) part of the schema. IT would seem if we can edit the schema programmatically (OSchema) we should be able to still add and drop indices via the Java API.

It seems like you can create the index using `OClass.createIndex`.  But I do not see a way to drop it. For now it seems like only `db.getMetadata.getIndexManager.dropIndex(name)` is possible.

**Comment by gianluca-valentini** [comment]
Hi,
I follow this question as I'm in the same situation. I need to migrate from 3.0 to 3.1 and I found some deprecated methods.
OIndexManager is now marked as deprecated, as is getIndexManager() from OMetadata but I use **ODatabasePool** to acquire an **ODatabaseSession**.

in my situation I have
```
 ODatabaseSession graph =...
OIndex idx = graph.getMetadata().getIndexManager().getIndex("indexName");
ORecordId id = (ORecordId) idx.get(key);
```

that are deprecated and I don't find anything (documentation too) that can help me to understand how to fix them

**Comment by andrii0lomakin** [comment]
Hi @gianluca-valentini , you need to use a mix of documents and simple select queries.
```
OCalss cls = schema.createClass("LookupTable");
cls.createProperty("key", OType.STRING);
cls.createProperty("value", ...);

//create an index on a key by any means which you prefer.

ODocument document = new ODocument("LookupTable");
document.setProperty("key", ..);
document.setProperty("value", ..);

graph.query("select * from LookupTable where key = ...");

```

**Reply by gianluca-valentini** [reply]
Thanks @laa, 
this means that the index is used autmaticaly even if I don't specify it in my query, isn't it?

What I should change the delete by index `ORecordId id = (ORecordId) idx.get(key); --> graph.delete(id);`?
Just for my understanding, this means that I can remove the OIndex usage in my code because it is used by Orient when I find items using query?

Thanks

**Reply by andrii0lomakin** [reply]
> this means that the index is used autmaticaly even if I don't specify it in my query, isn't it?

Yes, that is true.

> What I should change the delete by index ORecordId id = (ORecordId) idx.get(key); --> graph.delete(id);?

```
DELETE FROM (SELECT expand(value)  FROM LookupTable where key=...)
```

> Just for my understanding, this means that I can remove the OIndex usage in my code because it is used by Orient when I find items using query?

Yes
